src/sync/barrier.rs

```diff
@@ -56,57 +67,77 @@ impl Barrier {
         let me = ExecutionState::me();
 
         let mut state = self.state.borrow_mut();
-        trace!(leader=?state.leader, waiters=?state.waiters, "waiting on barrier {:p}", self);
-
-        // TODO The documentation for `Barrier` states that
-        // TODO    A single (arbitrary) thread will receive a `BarrierWaitResult` that returns true
-        // TODO    from `BarrierWaitResult::is_leader()` when returning from this function
-        // TODO Currently, the first waiter becomes the leader.  We should use Shuttle's nondeterministic
-        // TODO choice to generalize this so that _any_ participant could become the leader.
-        if state.leader.is_none() {
-            state.leader = Some(me);
-        }
+        let my_epoch = state.epoch;
+
+        trace!(waiters=?state.waiters, epoch=my_epoch, "waiting on barrier {:p}", self);
 
         // Update the barrier's clock with the clock of this thread
         ExecutionState::with(|s| {
             let clock = s.increment_clock();
             state.clock.update(clock);
         });
 
-        if state.waiters.len() + 1 < state.bound {
-            // Block current thread
-            assert!(state.waiters.insert(me)); // current thread shouldn't already be in the set
+        // The current thread shouldn't already be in `waiters`.
+        assert!(state.waiters.insert(me));
```

Comment:
I think I marginally preferred the existing code that didn't needlessly add the current thread to the waiters set, so that the invariant was that all waiters are blocked. But we should still assert here that `me` isn't already a waiter.

Reply:
>  It ends up having roughly the same number of special cases for me, just different ones.

What special cases for `me` do you mean? Assuming I kill `assert_eq!(t.blocked(), tid != me)` (which is really just for debugging), then the only use of `me` in the function is in the line you commented on here which I could just inline as:

```rust
assert!(state.waiters.insert(ExecutionState::me()));
```

---

> [...] the invariant was that all waiters are blocked.

Arguably, that still is an invariant. The barrier's `state` is never release while this isn't true (and every thread in `waiters` is always one that called `wait`).